op-node/sources/types.go

```diff
@@ -83,71 +87,175 @@ func (info *HeaderInfo) ReceiptHash() common.Hash {
 		}
 	}
+
+	info := HeaderInfo{
+		hash:        hdr.Hash,
+		parentHash:  hdr.ParentHash,
+		coinbase:    hdr.Coinbase,
+		root:        hdr.Root,
+		number:      uint64(hdr.Number),
+		time:        uint64(hdr.Time),
+		mixDigest:   hdr.MixDigest,
+		baseFee:     (*big.Int)(hdr.BaseFee),
+		txHash:      hdr.TxHash,
+		receiptHash: hdr.ReceiptHash,
+	}
 	return &info, nil
 }
 
-type rpcBlockCacheInfo struct {
+type rpcBlock struct {
+	rpcHeader
 	Transactions []*types.Transaction `json:"transactions"`
 }
 
-type rpcBlock struct {
-	header rpcHeader
-	extra  rpcBlockCacheInfo
+func (block *rpcBlock) verify() error {
+	if computed := block.computeBlockHash(); computed != block.Hash {
+		return fmt.Errorf("failed to verify block hash: computed %s but RPC said %s", computed, block.Hash)
+	}
+	if computed := types.DeriveSha(types.Transactions(block.Transactions), trie.NewStackTrie(nil)); block.TxHash != computed {
+		return fmt.Errorf("failed to verify transactions list: computed %s but RPC said %s", computed, block.TxHash)
+	}
+	return nil
 }
 
-func (block *rpcBlock) UnmarshalJSON(msg []byte) error {
-	if err := json.Unmarshal(msg, &block.header); err != nil {
-		return err
+func (block *rpcBlock) Info(trustCache bool, mustBePostMerge bool) (*HeaderInfo, types.Transactions, error) {
+	if mustBePostMerge {
+		if err := block.checkPostMerge(); err != nil {
+			return nil, nil, err
+		}
+	}
+	if !trustCache {
+		if err := block.verify(); err != nil {
+			return nil, nil, err
+		}
 	}
-	return json.Unmarshal(msg, &block.extra)
-}
 
-func (block *rpcBlock) Info(trustCache bool) (*HeaderInfo, types.Transactions, error) {
 	// verify the header data
-	info, err := block.header.Info(trustCache)
+	info, err := block.rpcHeader.Info(trustCache, mustBePostMerge)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to verify block from RPC: %v", err)
+		return nil, nil, fmt.Errorf("failed to verify block from RPC: %w", err)
 	}
 
-	if !trustCache { // verify the list of transactions matches the tx-root
-		hasher := trie.NewStackTrie(nil)
-		computed := types.DeriveSha(types.Transactions(block.extra.Transactions), hasher)
-		if expected := info.txHash; expected != computed {
-			return nil, nil, fmt.Errorf("failed to verify transactions list: expected transactions root %s but retrieved %s", expected, computed)
+	return info, block.Transactions, nil
+}
+
+func (block *rpcBlock) ExecutionPayload(trustCache bool) (*eth.ExecutionPayload, error) {
+	if err := block.checkPostMerge(); err != nil {
+		return nil, err
+	}
+	if !trustCache {
+		if err := block.verify(); err != nil {
+			return nil, err
 		}
 	}
-	return info, block.extra.Transactions, nil
+	var baseFee uint256.Int
+	baseFee.SetFromBig((*big.Int)(block.BaseFee))
+
+	// Unfortunately eth_getBlockByNumber either returns full transactions, or only tx-hashes.
+	// There is no option for encodes transactions.
```

Comment:
For L2 we can, and it's definitely an optimization that's worth looking at, but for L1 we can't change it. However, we use this primarily to just check if the derived block attributes data matches the L2 chain, so maybe there's a way to refactor it to only check the tx-root from an `ExecutionPayloadHeader` instead of full payload. That's worth a separate PR if we're refactoring derivation related code though, I'm ok with the way this is now, my previous L2 payload fetching code was worse :upside_down_face: